docs/reference/projectyml.rst

```diff
@@ -50,66 +50,122 @@ The code snippet below provides an example of a complete ``project.yml`` file:
+The :class:`~inmanta.module.ProjectPipConfig` section of the project.yml file offers options to configure this behaviour.
 
-Using pip config file
+pip.use-system-config
 """""""""""""""""""""
 
-By setting the ``use_config_file`` option of the pip section to ``True``, the project will use the pip config files.
+This option determines the isolation level of the project's pip config. When false, any pip config set on the system
+(e.g. through environment variables or pip config files) are ignored and pip will only look for packages
+in the index(es) defined in the project.yml, when true, pip will in addition look in the eventual index(es) defined on the system.
+
+Setting this to ``false`` is recommended during development both for portability (by making sure that only the pip
+config defined in the project.yml will be used regardless of the sytem's pip config) and for security (The isolation
+reduces the risk of dependency confusion attacks if the ``index-url`` option is set mindfully).
+
+Setting this to ``true`` will have the following consequences:
+
+- If no index is set in the project.yml file i.e. both ``index-url`` and ``extra-index-url`` are unset, then Pip's
+  default search behaviour will be used: environment variables, pip config files and then PyPi (in that order).
+
+- If ``index-url`` and/or ``extra-index-url`` are set, they will be used and any index defined in the system's environment
+  variables or pip config files will also be used (in that order) and passed to pip as extra indexes.
+
+- The ``PIP_PRE`` environment variable (if set) is no longer ignored and will be used to determine whether pre-release
+  versions are allowed when installing v2 modules or v1 modules' dependencies.
+
+Example scenario
+""""""""""""""""
+
+1) During development
+
+Using a single pip index isolated from any system config is the recommended approach. The ``pre=true`` option allows
+pip to use pre-release versions, e.g. when testing dev versions of modules published to the dev index. Here is an
+example of a dev config:
 
 .. code-block:: yaml
 
     pip:
-      use_config_file: True
+        index-url: https://devpi.example.com/dev/
+        extra-index-url: []
+        pre: true
+        use-system-config: false
 
-To specify the url of a pip repository, add the following to the pip config file of the ``inmanta`` user, located at ``/var/lib/inmanta/.config/pip/pip.conf``:
+2) In production
 
-.. code-block:: text
+Using a single pip index is still the recommended approach, and the use of pre-release versions should be disabled. Here is an
+example of a config suitable in production:
 
-  [global]
-  timeout = 60
-  index-url = <url of a python package repository>
+.. code-block:: yaml
 
+    pip:
+        index-url: https://devpi.example.com/stable/
+        extra-index-url: []
+        pre: false
+        use-system-config: true
```

Comment:
I find this one a bit confusing. 

I would also explain that for production scenario's, we expect all packages to be hosted in an internal, curated python repository (like nexus or devpi) and that the default pip config, which pull packages from `https://pypi.org/` is never used.

(this explanation may also belong elsewhere,  like e.g. https://docs.inmanta.com/inmanta-service-orchestrator/5/administrators/operational_procedures.html#project-release-for-production )

In production scenarios, where the system config is fully managed at the system level and secure, I would expect just `use-system-config: true`  or perhaps also a `extra-index-url:` for a project specific repo that is not managed at the system level.

In production scenarios, where the system config is not managed and secure, I would not expect use-system-config and to set `index-url:` to the secure internal repo.

```
  pip:
        index-url: https://my_repo.secure.company.com/repository/inmanta-production
        extra-index-url: [https://my_repo.secure.company.com/repository/project-alpha-production]
        pre: false
        use-system-config: false
 ```